KAFKA-19744: Move restore time calculation to ChangelogMetadata

Nikita-Shupletsov opened the pull request:
- Move restore time calculation to ChangelogMetadata.
- Introduced a new interface to propagate the calculated value to the
stores to avoid modifications in the public interface.

Reviewers: Matthias J. Sax <matthias@confluent.io>

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredInternalProcessorContext.java:
```diff
@@ -0,0 +1,312 @@
+
+        @Override
+        public void restore(final byte[] key, final byte[] value) {
+            StreamsMetricsImpl.maybeMeasureLatency(() -> wrapped.restore(key, value), time, restoreSensor);
```
The call to `restore()` does restore one "batch" of data, ie, whatever a single `poll()` call returned. So this is not the latency from "begin restore" to "complete restore".

We could maybe use `StateRestoreListener` callback, but also not sure if that's the best way?

I don't think we could use `maybeMeasureLatency()` helper method at all, but would need to get a ts when restore stars, and also when restore completes, and call `sensor.record(...)` directly?

This would go into StoreChangelogReader? Or we use existing `Task#recordRestoration()` method (or add similar overloads). We need to break it out per-store 🤔

Nikita-Shupletsov replied:
completely reworked the PR. please let me know what you think.

mjsax reviewed (commented):
This PR updates `MeteredKeyValueStore` and `MeteredSessionStore`, but there is also `MeteredWindowStore`.

There is also `MeteredVersionStore` which does not use the sensor -- so I assume, we incorrectly don't record the restore metric for versioned stores? -- We should also close this gap.

Nikita-Shupletsov committed:
KAFKA-19744: Move restore time calculation to ChangelogMetadata.
Intorduced a new interface to propogate the calculated value to the stores to avoid modifications in the public interface.

Nikita-Shupletsov commented:
> This PR updates `MeteredKeyValueStore` and `MeteredSessionStore`, but there is also `MeteredWindowStore`.
> 
> There is also `MeteredVersionStore` which does not use the sensor -- so I assume, we incorrectly don't record the restore metric for versioned stores? -- We should also close this gap.

thanks for catching that.
I took a look. MeteredTimestampedKeyValueStore, MeteredTimestampedKeyValueStore, and MeteredVersionedKeyValueStoreInternal extend MeteredKeyValueStore. and MeteredTimestampedWindowStore extends MeteredWindowStore

so the only missed metered store was MeteredWindowStore. at least I wasn't able to find anything else

mjsax commented:
Seems a test is breaking: `FAILED ❌ MeteredVersionedKeyValueStoreTest > shouldRecordMetricsOnInit()`

But given the name of the test, it might actually be void, and we should remove the entire test, as we actually don't want to record any metrics in `init()`?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -188,6 +191,10 @@ List<ConsumerRecord<byte[], byte[]>> bufferedRecords() {
             return bufferedLimitIndex;
         }
+
+        long calculateRestoreTime(final Time time) {
```
Why do we pass `Time` instead of `long restoreCompletionTimeNs` ? (Basic API design suggest, to pass the minimum what you need.)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredStateStore.java:
```diff
@@ -0,0 +1,24 @@
+
+import org.apache.kafka.streams.processor.StateStore;
+
+public interface MeteredStateStore extends StateStore {
```
Seems unnecessary (even if not harmful) to let it `extends StateStore` ?

Nikita-Shupletsov replied:
I extended StateStore for the ease of mocking. will fix that

Nikita-Shupletsov committed:
Removed the obsolete test.

Nikita-Shupletsov commented:
> Seems a test is breaking: `FAILED ❌ MeteredVersionedKeyValueStoreTest > shouldRecordMetricsOnInit()`
> 
> But given the name of the test, it might actually be void, and we should remove the entire test, as we actually don't want to record any metrics in `init()`?

yes, you are right. sorry, idk how I missed that test. deleted

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStoreTest.java:
```diff
@@ -210,12 +210,15 @@ public void testMetrics() {
 
         init();
 
+        final long restoreTimeNs = 1000;
```
```suggestion
        final long restoreTimeNs = 1000L;
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredSessionStoreTest.java:
```diff
@@ -458,10 +458,13 @@ public void shouldNotFindExpiredSessionRangeInBackwardOrderFromStore() {
         setUp();
         init();
 
+        final long restoreTimeNs = 1000;
```
```suggestion
        final long restoreTimeNs = 1000L;
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredWindowStoreTest.java:
```diff
@@ -209,6 +209,22 @@ public void testMetrics() {
+
+        store.init(context, store);
+
+        final long restoreTimeNs = 1000;
```
```suggestion
        final long restoreTimeNs = 1000L;
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredWindowStoreTest.java:
```diff
@@ -209,6 +209,22 @@ public void testMetrics() {
+        // it suffices to verify one restore metric since all restore metrics are recorded by the same sensor
+        // and the sensor is tested elsewhere
+        final KafkaMetric metric = metric("restore-rate");
+        assertThat((Double) metric.metricValue(), greaterThan(0.0));
```
With the old code, it was not clear what latency we would measure, to we use `greaterThan`, but know we know that we should record exactly `1000`. Should we use `is(1000L)` instead?

Seems the existing tests for the other store use `assertTrue(...)` should we also update them to use `assertThat` ?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreChangelogReaderTest.java:
```diff
@@ -1364,6 +1366,29 @@ public void shouldNotThrowOnUnknownRevokedPartition() {
+
+        changelogReader.register(tp, activeStateManager);
+
+        changelogReader.restore(Collections.singletonMap(taskId, mock(Task.class)));
```
It's not totally clear to me, why this call ensure that restore completes, and we hit the right code path?

Should we also add a negative test, `shouldNotRecordRestoreTimeIfRestoreDoesNotComplete`?

Nikita-Shupletsov replied:
I updated the test. hopefully, it's a bit clearer what it's trying to test

mjsax reviewed (commented):
Thanks for updating the PR. I think this works. Just a a couple of nits and minor things.

mjsax commented:
One more thought: Can we extend `RestoreIntegrationTest`, using `TrackingStateRestoreListener` to also establish a bound on restore latency, and compare our bound to the collected metric, ie, `bound - 1 sec < metric-value< bound + 1 sec` (to allow for some jitter).

Nikita-Shupletsov committed:
Update streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStoreTest.java

Co-authored-by: Matthias J. Sax <mjsax@apache.org>

Nikita-Shupletsov committed:
Update streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredSessionStoreTest.java

Co-authored-by: Matthias J. Sax <mjsax@apache.org>

Nikita-Shupletsov committed:
Update streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredWindowStoreTest.java

Co-authored-by: Matthias J. Sax <mjsax@apache.org>

Nikita-Shupletsov committed:
Addressed comments.

mjsax reviewed (approved):
Thanks for the PR. LGTM. (It's a tricky one; hope we did not miss anything.)
